src/main/java/com/apicatalog/jsonld/ModifiableJsonArray.java

```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Override
+    public JsonNumber getJsonNumber(int index) {
+        return (JsonNumber) valueList.get(index);
+    }
+
+    @Override
+    public JsonString getJsonString(int index) {
+        return (JsonString) valueList.get(index);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends JsonValue> List<T> getValuesAs(Class<T> clazz) {
+        return (List<T>) valueList;
+    }
+
+    @Override
+    public String getString(int index) {
+        return getJsonString(index).getString();
+    }
+
+    @Override
+    public String getString(int index, String defaultValue) {
+        try {
+            return getString(index);
+        } catch (Exception e) {
+            return defaultValue;
+        }
+    }
+
+    @Override
+    public int getInt(int index) {
+        return getJsonNumber(index).intValue();
+    }
+
+    @Override
+    public int getInt(int index, int defaultValue) {
+        try {
+            return getInt(index);
+        } catch (Exception e) {
+            return defaultValue;
+        }
+    }
+
+    @Override
+    public boolean getBoolean(int index) {
+        JsonValue jsonValue = get(index);
+        if (jsonValue == JsonValue.TRUE) {
+            return true;
+        } else if (jsonValue == JsonValue.FALSE) {
+            return false;
+        } else {
+            throw new ClassCastException();
+        }
+    }
+
+    @Override
+    public boolean getBoolean(int index, boolean defaultValue) {
+        try {
+            return getBoolean(index);
+        } catch (Exception e) {
+            return defaultValue;
+        }
+    }
+
+    @Override
+    public boolean isNull(int index) {
+        return valueList.get(index).equals(JsonValue.NULL);
+    }
+
+    @Override
+    public ValueType getValueType() {
+        return ValueType.ARRAY;
+    }
+
+    @Override
+    public JsonValue get(int index) {
+        return valueList.get(index);
+    }
+
+    public boolean add(JsonValue value) {
+        if (filter != null) {
+            filter.put(value.hashCode());
+        }
+        return valueList.add(value);
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if(valueList.isEmpty()) return false;
+        if(o == null) return false;
+
+        if (filter == null && valueList.size() > 100) {
+            filter = BloomFilter.create(
+                    Funnels.integerFunnel(),
+                    100000,
+                    0.01);
+            for (JsonValue value : valueList) {
+                filter.put(value.hashCode());
+            }
+        }
+        if (filter != null) {
+            boolean mightContain = filter.mightContain(o.hashCode());
+            if (!mightContain) {
+                return false;
+            }
+        }
+
+        int oHashCode = o.hashCode();
+        if(valueList.size() == 1){
+            JsonValue jsonValue = valueList.get(0);
+            if(oHashCode == jsonValue.hashCode()){
```

Comment:
Please note  that  `a.hashCode()  == b..hashCode()` could be true even if those objects are different.  [an explanation](https://stackoverflow.com/questions/29714448/why-cant-i-just-compare-the-hashcode-of-two-objects-in-order-to-find-out-if-the)
Our of my curriosity, why is the `equals` method so slow?

Reply:
The various objects will cache their hash codes after they are calculated. The equals method can't really cache anything. This makes checking if the hash code is equal a lot faster. Two objects with different hash codes should never be equal, but of course there is nothing stopping a hash collision. Hash collisions are fairly rare though, especially since we are not truncating the hash code in any way. They do occur though, which is why I confirm with .equals(...) as a safe guard. 

The general check of if the `reference` element already exists in the array seems to be a way of handling an edge case, since it's never the case while processing the 600 000 test file I'm using.